### PR TITLE
Mixer source

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Last refactoring step before really making rendering of effects work.

Extract an AudioGraph::Source subclass from Mixer, which will be adapted with EffectStage.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: libraries/lib-sample-track/Mix.h [hunk 1]
```diff
@@ -271,5 +253,86 @@ class SAMPLE_TRACK_API Mixer {
...
+ reference to its base class.
+ */
+class MixerSource final : public AudioGraph::Source {
+public:
+   using TimesAndSpeed = MixerOptions::TimesAndSpeed;
+
+   /*!
+    @pre `pTimesAndSpeed != nullptr`
+    */
+   MixerSource(const SampleTrack &leader, size_t i,
```
Comment (crsib) [inline libraries/lib-sample-track/Mix.h]: Generally I'm not a fan of such a long arguments lists,

File: libraries/lib-sample-track/Mix.cpp [hunk 2]
```diff
@@ -210,8 +210,10 @@ double ComputeWarpFactor(const Envelope &env, double t0, double t1)
 }
 
 size_t MixerSource::MixVariableRates(
-   size_t ii, const size_t maxOut, float &floatBuffer)
+   size_t j, const size_t maxOut, float &floatBuffer)
```
Comment (crsib) [inline libraries/lib-sample-track/Mix.cpp]: `j` is as bad as `ii` to be fair.

File: libraries/lib-sample-track/Mix.cpp [hunk 3]
```diff
@@ -210,8 +210,10 @@ double ComputeWarpFactor(const Envelope &env, double t0, double t1)
...
 {
+   const auto ii = i + j;
```
Comment (crsib) [inline libraries/lib-sample-track/Mix.cpp]: In the context of this review I have no clue what is i, j and ii.

File: libraries/lib-sample-track/Mix.cpp [hunk 4]
```diff
@@ -571,13 +589,7 @@ size_t Mixer::Process(const size_t maxToProcess)
...
       auto increment = finally([&]{ i += nInChannels; });
 
-      MixerSource source{ *leader, i, mRate,
-         mResampleParameters.mbVariableRates, mEnvelope, mMayThrow
-         , mTimesAndSpeed
-         , mInputTrack, mSamplePos, mSampleQueue, mQueueStart
-         , mQueueLen, mResample, mEnvValues
-      };
-      auto oResult = source.Acquire(mFloatBuffers, maxToProcess);
+      auto oResult = pSource++ ->Acquire(mFloatBuffers, maxToProcess);
```
Comment (crsib) [inline libraries/lib-sample-track/Mix.cpp]: Looks quite awkward to be fair

**Review by crsib — APPROVED** [review]

File: libraries/lib-sample-track/Mix.h (hunk 1 above)
Comment (Paul-Licameli) [reply libraries/lib-sample-track/Mix.h]: Of course not, and the purpose of later commits is to shorten them after that rough separation

File: libraries/lib-sample-track/Mix.cpp [hunk 5]
```diff
@@ -34,33 +34,23 @@
...
-   {
-      wxASSERT(false);
-      maxSpeed = 0;
-   }
-   if (minSpeed > maxSpeed)
-   {
-      wxASSERT(false);
-      std::swap(minSpeed, maxSpeed);
-   }
+   assert(min >= 0);
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-sample-track/Mix.cpp]: minSpeed and maxSpeed were reassigned, but not the arguments min, max.  I preserve the assertions, and make explicit the precondition in the header, while also making this struct an immutable -- therefore, it doesn't need private members or friends.  The constructor establishes the invariant property of the structure which won't be violated thereafter, barring dirty tricks with casts.

**Pull request merged**